jdbc/src/main/java/nextstep/jdbc/JdbcTemplate.java

```diff
@@ -1,9 +1,62 @@
 package nextstep.jdbc;
 
+import java.sql.Connection;
+import java.sql.PreparedStatement;
+import java.sql.SQLException;
+import java.util.List;
+import javax.sql.DataSource;
+
 import org.slf4j.Logger;
 import org.slf4j.LoggerFactory;
 
 public class JdbcTemplate {
 
     private static final Logger log = LoggerFactory.getLogger(JdbcTemplate.class);
+
+    protected final DataSource dataSource;
+
+    public JdbcTemplate(DataSource dataSource) {
+        this.dataSource = dataSource;
+    }
+
+    public void update(String sql, Object... params) {
+        execute(sql, PreparedStatement::executeUpdate, params);
+    }
+
+    public <T> List<T> queryForList(String sql, RowMapper<T> rowMapper, Object... params) {
+        final StatementCallback<List<T>> callback =
+                statement -> new ResultSetExtractor<>(rowMapper).extractData(statement.executeQuery());
+        return execute(sql, callback, params);
+    }
+
+    public <T> T queryForObject(String sql, RowMapper<T> rowMapper, Object... params) {
+        final List<T> results = queryForList(sql, rowMapper, params);
+        if (results.isEmpty()) {
+            throw new IncorrectResultSizeDataAccessException("파라미터에 해당하는 엔티티를 찾을 수 없습니다.");
+        }
+
+        if (results.size() > 1) {
+            throw new IncorrectResultSizeDataAccessException("파라미터와 일치하는 엔티티가 2개 이상입니다.");
+        }
+
+        return results.get(0);
```

Comment:
```suggestion
        if (results.size() == 1) {
            return results.get(0);
        }
        throw new IncorrectResultSizeDataAccessException("예외 메시지");
```

이렇게 작성하면 더 깔끔할 것 같아요 ~!

Reply:
```suggestion
        if (isValidSize()) {
            return results.get(0);
        }
        throw new IncorrectResultSizeDataAccessException("예외 메시지");
        
    private boolean isValidSize() {
        return results.size() == 1;
    }
```

그리고 if 문의 조건을 메소드 분리하면 추상화 수준을 높게 가져갈 수 있어요 !!

Reply:
DB에 존재하지 않는 엔티티를 검색하려고 한 경우와 DB에 중복된 엔티티가 들어있는 경우의 예외를 다르게 표시하려고 구분해 둔 예외 처리였는데, 어차피 둘이 같은 예외를 사용하므로 다니 말대로 하나로 합치는 것도 좋은 방법인 것 같네요.
하나로 합치는 방식으로 수정했습니다!